### PR TITLE
Code1

### PR DESCRIPTION
**Pull request by Priyanka-droid**
A menu driven command line dart program that provides the following menu options:
1.Add User details.
2.Display User details.
3.Delete User details
4.Save User details.
5.Exit

for option 1 details should be:
1. Full Name
2. Roll Number
3. Age
4. Address
5. Set fo courses( 4 out of 6 required)

for option 2 display details in sorted order based on attribute and order

for option 3 delete detail by roll number

for option 4 save the data to disk

for option 5 exit app and before exit ask if want to save details

### CODE REVIEW

File: dart_application_2/test/dart_application_2_test.dart [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import 'package:dart_application_2/dart_application_2.dart';
+import 'package:test/test.dart';
+
+void main() {
+  test('calculate', () {
+    expect(calculate(), 42);
+  });
+}
```
Comment (ajayvenkat10) [inline dart_application_2/test/dart_application_2_test.dart]: Write unit test cases for your functions that solve business logics in dart
Will be easier for me to download the project and run the tests to see if it goes right
Remember to think of all use cases, happy flow, errors, exceptions, edge cases etc

File: dart_application_2/bin/college_app.dart [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import 'validations_util.dart';
+
+class CollegeApp {
+  static void startApp() {
+    String optionString;
+    int option;
+    bool exceptionFlag;
+    // list of students
+    List<Student> studentList = [];
```
Comment (ajayvenkat10) [inline dart_application_2/bin/college_app.dart:17]: Try with SplayTreeSet now

File: dart_application_2/bin/college_app.dart [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    // // if file does not exist create the file with empty data
+    if (!fileExist) {
+      file.writeAsStringSync(jsonEncode([]), mode: FileMode.write);
+    }
+    // read content of file and save to the list
+
+    String contents = file.readAsStringSync();
+    var jsonContent = jsonDecode(contents);
+    studentList =
+        List<Student>.from(jsonContent.map((model) => Student.fromJson(model)));
```
Comment (ajayvenkat10) [inline dart_application_2/bin/college_app.dart]: File I/O should be part of a different class 
Call it LocalStorage / FileIO etc
Separate read and write functions

File: dart_application_2/bin/college_app.dart (hunk 2 above)
Comment (ajayvenkat10) [inline dart_application_2/bin/college_app.dart:17]: Lets implement singleton pattern 
Read about singleton pattern and singleton classes in dart 
Create a singleton class , for example : ApplicationInstance 
And when the singleton instance is created call the read function to read from disk storage 
Singleton - Only one instance of the class can exist per session 
Why singleton ? 
As only one instance can be created of it, it helps in performing the task of reading from the disk which should be a one time task per session and also you can declare your set/list of Student objects there as that is an object used throughout the application and is easier to access. 
Think about what all objects and functions can be part of this class

File: dart_application_2/bin/college_app.dart [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        optionString = stdin.readLineSync()!;
+        exceptionFlag = ValidationsUtil.validateOption(optionString, 1, 5);
+      } while (exceptionFlag);
+      option = int.parse(optionString);
+
+      // perform the task for given option and save the reult in list
+      studentList = Task.performTask(option, studentList, file);
+
+      // sort the list by FullName and then Roll Number
+      studentList.sort();
```
Comment (ajayvenkat10) [inline dart_application_2/bin/college_app.dart:43]: SplayTreeSet could help avoid sorting after each input

File: dart_application_2/bin/task.dart [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        state,
+        rollString,
+        courseString;
+    int age, rollNumber, streetNumber;
+    bool exceptionFlag;
+    do {
+      exceptionFlag = false;
+      print("enter name of student");
+      fullName = stdin.readLineSync()!;
+      exceptionFlag = ValidationsUtil.validateString(fullName);
```
Comment (ajayvenkat10) [inline dart_application_2/bin/task.dart]: Can avoid exceptionFlag parameter 
Just need to put ValidationsUtil.validate() method as the condition in while as it returns a boolean
Just writing it here 
Can follow the same for the below ones too

File: dart_application_2/bin/task.dart [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  static _displayDetailsOption(List<Student> studentList) {
+    studentList = SortItems.sortItems(studentList);
+    studentList.forEach((student) {
+      print("${student.fullName}" +
+          " ${student.rollNumber}  ${student.age})" +
+          " ${student.address.streetNumber}" +
+          "${student.address.city} ${student.address.state}" +
+          "${student.courseSet}");
+    });
+  }
```
Comment (ajayvenkat10) [inline dart_application_2/bin/task.dart]: You have to display in the table format as given in the problem statement

File: dart_application_2/bin/task.dart [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    studentList.removeAt(indexRollNumber);
+    return studentList;
+  }
+
+  /**
+   * save details from in-memory to disk
+   */
+  static _saveDetailsOption(List<Student> studentList, File file) {
+    String jsonTags = jsonEncode(studentList);
+    file.writeAsStringSync(jsonTags, mode: FileMode.write);
```
Comment (ajayvenkat10) [inline dart_application_2/bin/task.dart]: FileIO should be part of FileIO class

File: dart_application_2/bin/task.dart [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+      do {
+        exceptionFlag = false;
+        print("enter roll number to be deleted");
+        rollNumberString = stdin.readLineSync()!;
+        exceptionFlag = ValidationsUtil.validateNumber(rollNumberString);
+      } while (exceptionFlag);
+      rollNumber = int.parse(rollNumberString);
+      indexRollNumber = studentList
+          .indexWhere(((student) => student.rollNumber == rollNumber));
+    } while (indexRollNumber == -1);
```
Comment (ajayvenkat10) [inline dart_application_2/bin/task.dart:168]: What is this while and its condition for ? 
You've handled the validation in the inner while 
Is this while required ? 
Also not required to do indexWhere and removeAt 
Checkout list.removeWhere, it does what you need in one step

File: dart_application_2/bin/task.dart [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+   * chose options before closing app:
+   * 4: save and close
+   * 5: close
+   */
+  static _exitApp(List<Student> studentList, File file) {
+    bool exceptionFlag;
+    String optionString;
+    do {
+      exceptionFlag = false;
+      print("enter 4 to save, 5 to exit");
```
Comment (ajayvenkat10) [inline dart_application_2/bin/task.dart]: Instead take a yes or no from the user
As the user if he/she wants to save the details before exiting 
If yes -> save and exit, else just exit

File: dart_application_2/bin/student.dart [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    if (this.fullName.compareTo(otherStudent.fullName) == 1) {
+      return 1;
+    }
+    if (this.fullName.compareTo(otherStudent.fullName) == 0) {
+      if (this.rollNumber < otherStudent.rollNumber) return -1;
+      if (this.rollNumber >= otherStudent.rollNumber) return 1;
+    }
+    return 0;
```
Comment (ajayvenkat10) [inline dart_application_2/bin/student.dart]: Easier way to do:
1. Store the compareTo result in a variable say 'compareResult'
2. `return compareResult == 0 ? this.rollNumber.compareTo(other.rollNumber) : compareResult;`

File: dart_application_2/bin/exceptions.dart [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  static String errMsg() => "Value entered is out of range";
+}
+
+class AlphabetException implements Exception {
+  static String errMsg() => "Name should only contain alphabets";
+}
+
+class CaseException implements Exception {
+  static String errMsg() => "First letter of name should be capital";
+}
```
Comment (ajayvenkat10) [inline dart_application_2/bin/exceptions.dart]: Not a bad way but can be better.
By using Inheritance here, its easier to create each exception.

For example:

```
abstract class CollegeAppException implements Exception{

  final String errorMessage;

  CollegeAppException(this.errorMessage);

  @override
  String toString() {
    return errorMessage;
  }
}

class IllegalTypeException extends CollegeAppException{
  IllegalTypeException() : super('Illegal Type Exception');
}

class NumberException extends CollegeAppException{
  NumberException() : super('LegalException Exception');
}
```

And usage: 

```
try{
    throw IllegalTypeException();
  } on CollegeAppException catch(exception,stackTrace){
    print(exception.toString());
    print(stackTrace);
  }
```

This way you can avoid catching each of them separately and catch the base class pointer

File: dart_application_2/bin/validations_util.dart [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+   * 3. Natural Number
+   * 4. Within given range
+   */
+  static bool validateOption(String option, int rangeStart, int rangeEnd) {
+    bool exceptionFlag = false;
+    try {
+      if (option.isEmpty) throw new EmptyStringException();
+      if (!_isNumeric(option)) throw new NonNumericException();
+      if (_nonWhole(option)) throw new NonWholeException();
+      if (!_inRange(option, rangeStart, rangeEnd))
```
Comment (ajayvenkat10) [inline dart_application_2/bin/validations_util.dart]: Pass int.parse(option) here cause you've already validated for it being a number above 
So by the time it reaches here it will be a number for sure

File: dart_application_2/bin/task.dart [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+      rollString = stdin.readLineSync()!;
+      exceptionFlag = ValidationsUtil.validateNumber(rollString);
+     //         TODO: check if the student corresponding to this roll number already exists
+    } while (exceptionFlag);
+    
+    rollNumber = int.parse(rollString);
+
+    Set<Course> courseSet = {};
+
+    while (courseSet.length < 4) {
```
Comment (ajayvenkat10) [inline dart_application_2/bin/task.dart]: What is 4 ? 
Define as a constant

File: dart_application_2/bin/task.dart [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+      exceptionFlag = false;
+      print("chose 4 courses from 0,1,2,3,4,5");
+      courseString = stdin.readLineSync()!;
+      exceptionFlag = ValidationsUtil.validateOption(courseString, 0, 5);
```
Comment (ajayvenkat10) [inline dart_application_2/bin/task.dart]: What are 0 and 5 ? 
Define as constants

**Review by ajayvenkat10 — CHANGES_REQUESTED** [review]

File: dart_application_2/bin/task.dart (hunk 8 above)
Comment (Priyanka-droid) [reply dart_application_2/bin/task.dart:168]: The inner while only checks if the given roll number is passing the validation criteria of roll number, it does not check if that roll number exist in the list of students, the indexWhere checks if that roll number exist or not in the list, if that index is not in the list then another input for roll number will be asked again and validated again. The loop exits only when a roll number present in the list is entered and then details of that corresponding roll number is deleted using removeWhere.
